<a href="https://colab.research.google.com/github/taycurran/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/CURRAN_112_Make_Features_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200> 

# Assignment:

- Replicate the lesson code.

 - This means that if you haven't followed along already, type out the things that we did in class. Forcing your fingers to hit each key will help you internalize the syntax of what we're doing. Make sure you understand each line of code that you're writing, google things that you don't fully understand.
 - [Lambda Learning Method for DS - By Ryan Herr](https://docs.google.com/document/d/1ubOw9B3Hfip27hF2ZFnW3a3z9xAgrUDRReOEo-FHCVs/edit?usp=sharing)
- Convert the `term` column from string to integer.
- Make a column named `loan_status_is_great`. It should contain the integer 1 if `loan_status` is "Current" or "Fully Paid." Else it should contain the integer 0.
- Make `last_pymnt_d_month` and `last_pymnt_d_year` columns.

In [0]:
##### Begin Working Here #####

In [0]:
import pandas as pd

In [0]:
df = pd.read_csv('https://raw.githubusercontent.com/ryanleeallred/datasets/master/Ames%20Housing%20Data/train.csv')

df.info()

In [6]:
df.head()

Id  MSSubClass MSZoning  ...  SaleType  SaleCondition SalePrice
0   1          60       RL  ...        WD         Normal    208500
1   2          20       RL  ...        WD         Normal    181500
2   3          60       RL  ...        WD         Normal    223500
3   4          70       RL  ...        WD        Abnorml    140000
4   5          60       RL  ...        WD         Normal    250000

[5 rows x 81 columns]

In [7]:
df.shape

(1460, 81)

^So we can see: 
*   81 Attributes or Features
*   1460 Observations or Instances



In [8]:
pd.set_option('display.max_rows', 100)

df.dtypes

Id                 int64
MSSubClass         int64
MSZoning          object
LotFrontage      float64
LotArea            int64
Street            object
Alley             object
LotShape          object
LandContour       object
Utilities         object
LotConfig         object
LandSlope         object
Neighborhood      object
Condition1        object
Condition2        object
BldgType          object
HouseStyle        object
OverallQual        int64
OverallCond        int64
YearBuilt          int64
YearRemodAdd       int64
RoofStyle         object
RoofMatl          object
Exterior1st       object
Exterior2nd       object
MasVnrType        object
MasVnrArea       float64
ExterQual         object
ExterCond         object
Foundation        object
BsmtQual          object
BsmtCond          object
BsmtExposure      object
BsmtFinType1      object
BsmtFinSF1         int64
BsmtFinType2      object
BsmtFinSF2         int64
BsmtUnfSF          int64
TotalBsmtSF        int64
Heating           object


In [9]:
df['BedroomAbvGr'].head(10)
# Looks like integers.

0    3
1    3
2    3
3    3
4    4
5    1
6    3
7    3
8    2
9    2
Name: BedroomAbvGr, dtype: int64

In [11]:
# Looking at first ten rows of 'LotFrontage'
df['LotFrontage'].head(10)
# Clearly floats...

0    65.0
1    80.0
2    68.0
3    60.0
4    84.0
5    85.0
6    75.0
7     NaN
8    51.0
9    50.0
Name: LotFrontage, dtype: float64

In [12]:
df['LotFrontage'].value_counts(dropna=False)
# We observe Nan is most common value in this column.
# We observe round number bias.

NaN      259
60.0     143
70.0      70
80.0      69
50.0      57
        ... 
106.0      1
38.0       1
138.0      1
140.0      1
137.0      1
Name: LotFrontage, Length: 111, dtype: int64

In [13]:
df['LotFrontage'].value_counts(dropna=False).head()

NaN     259
60.0    143
70.0     70
80.0     69
50.0     57
Name: LotFrontage, dtype: int64

In [15]:
import numpy as np

type(np.NaN)

# ?- Why do we have to use numpy to examine this -?

float

In [17]:
small_df = df[['TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'SalePrice']].copy()

small_df.head()

TotalBsmtSF  1stFlrSF  2ndFlrSF  SalePrice
0          856       856       854     208500
1         1262      1262         0     181500
2          920       920       866     223500
3          756       961       756     140000
4         1145      1145      1053     250000

In [18]:
# Now we want to add the three SF columns to arrive at total SquareFootage
# We will create a new column for the small_df.

small_df['TotalSquareFootage'] = small_df['TotalBsmtSF'] + small_df['1stFlrSF'] + small_df['2ndFlrSF']

small_df.head()

TotalBsmtSF  1stFlrSF  2ndFlrSF  SalePrice  TotalSquareFootage
0          856       856       854     208500                2566
1         1262      1262         0     181500                2524
2          920       920       866     223500                2706
3          756       961       756     140000                2473
4         1145      1145      1053     250000                3343

In [19]:
# Now we want to create a new column, 'PricePerSqFt' through division.

small_df['PricePerSqFt'] = small_df['SalePrice'] / small_df['TotalSquareFootage']

small_df.head()

TotalBsmtSF  1stFlrSF  2ndFlrSF  SalePrice  TotalSquareFootage  PricePerSqFt
0          856       856       854     208500                2566     81.254871
1         1262      1262         0     181500                2524     71.909667
2          920       920       866     223500                2706     82.594235
3          756       961       756     140000                2473     56.611403
4         1145      1145      1053     250000                3343     74.783129

##NEW DATASET

In [20]:
!wget https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip

--2019-12-03 20:14:47--  https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip
Resolving resources.lendingclub.com (resources.lendingclub.com)... 64.48.1.20
Connecting to resources.lendingclub.com (resources.lendingclub.com)|64.48.1.20|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/zip]
Saving to: ‘LoanStats_2018Q4.csv.zip’

LoanStats_2018Q4.cs     [       <=>          ]  21.69M  1.35MB/s    in 16s     

2019-12-03 20:15:04 (1.33 MB/s) - ‘LoanStats_2018Q4.csv.zip’ saved [22744357]



In [21]:
!unzip LoanStats_2018Q4.csv.zip

Archive:  LoanStats_2018Q4.csv.zip
  inflating: LoanStats_2018Q4.csv    


In [24]:
!head LoanStats_2018Q4.csv

Notes offered by Prospectus (https://www.lendingclub.com/info/prospectus.action)
"id","member_id","loan_amnt","funded_amnt","funded_amnt_inv","term","int_rate","installment","grade","sub_grade","emp_title","emp_length","home_ownership","annual_inc","verification_status","issue_d","loan_status","pymnt_plan","url","desc","purpose","title","zip_code","addr_state","dti","delinq_2yrs","earliest_cr_line","inq_last_6mths","mths_since_last_delinq","mths_since_last_record","open_acc","pub_rec","revol_bal","revol_util","total_acc","initial_list_status","out_prncp","out_prncp_inv","total_pymnt","total_pymnt_inv","total_rec_prncp","total_rec_int","total_rec_late_fee","recoveries","collection_recovery_fee","last_pymnt_d","last_pymnt_amnt","next_pymnt_d","last_credit_pull_d","collections_12_mths_ex_med","mths_since_last_major_derog","policy_code","application_type","annual_inc_joint","dti_joint","verification_status_joint","acc_now_delinq","tot_coll_amt","tot_cur_bal","open_acc_6m","open_act_il","op

In [0]:
# Through surface examination we see that there is line above the header
# and two footer lines.

In [29]:
df = pd.read_csv('LoanStats_2018Q4.csv', header=1, skipfooter=2)
df.head()

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support skipfooter; you can avoid this warning by specifying engine='python'.
  """Entry point for launching an IPython kernel.


id  member_id  ...  settlement_percentage  settlement_term
0 NaN        NaN  ...                    NaN              NaN
1 NaN        NaN  ...                    NaN              NaN
2 NaN        NaN  ...                    NaN              NaN
3 NaN        NaN  ...                    NaN              NaN
4 NaN        NaN  ...                    NaN              NaN

[5 rows x 144 columns]

In [31]:
df.shape
# SHAPE =
# 128412 observations
# 144 attributes

(128412, 144)

In [32]:
df.isnull().sum().sort_values()

inq_fi                                             0
delinq_amnt                                        0
chargeoff_within_12_mths                           0
acc_open_past_24mths                               0
inq_last_12m                                       0
                                               ...  
orig_projected_additional_accrued_interest    128249
url                                           128412
member_id                                     128412
desc                                          128412
id                                            128412
Length: 144, dtype: int64

In [0]:
# We want to drop the 100% blank columns

df = df.drop(['url', 'member_id', 'desc', 'id'], axis=1)

In [37]:
cell = df['int_rate'][2]
type(cell)
# This column should not be a string.

str

In [38]:
# We need to go about cleaning this column.

float(cell.strip().strip('%'))

11.31

In [0]:
# Create a function for the .apply method

def clean_int(cell):
  return float(cell.strip().strip('%'))

In [0]:
# ^Ok so we have managed to process one cell... now to apply to column.
# We can do this using the pd.apply() method.

df['int_rate_cleaned'] = df['int_rate'].apply(clean_int)

# Now we replace our working output column to the original int_rate column

df['int_rate'] = df['int_rate_cleaned']

In [43]:
# Check type

type(df['int_rate'][2])

numpy.float64

In [0]:
# Drop working colum.

df.drop('int_rate_cleaned', axis=1, inplace=True)

In [49]:
df.head()

loan_amnt  funded_amnt  ...  settlement_percentage settlement_term
0      20000        20000  ...                    NaN             NaN
1      25000        25000  ...                    NaN             NaN
2      26500        26500  ...                    NaN             NaN
3      10400        10400  ...                    NaN             NaN
4      35000        35000  ...                    NaN             NaN

[5 rows x 140 columns]

In [51]:
df['emp_title'].value_counts(dropna=False)[:20]

NaN                   20947
Teacher                2090
Manager                1773
Registered Nurse        952
Driver                  924
RN                      726
Supervisor              697
Sales                   580
Project Manager         526
General Manager         523
Office Manager          521
Owner                   420
Director                402
Operations Manager      387
Truck Driver            387
Nurse                   326
Engineer                325
Sales Manager           304
manager                 301
Supervisor              270
Name: emp_title, dtype: int64

In [57]:
# We want to look at how many unique employment titles exsist in this column.

print(f"Unique Values: {len(df['emp_title'].value_counts())}")
print(f"Total Observations: {len(df['emp_title'])}")

Unique Values: 43892
Total Observations: 128412


In [58]:
# Number of null cases for emp_titile.
df['emp_title'].isnull().sum()

20947

In [0]:
# An easy way to approach this is to create a list of example'problem' inputs.

examples = ['owner', 'Supervisor', 'Project Manager', np.NaN]

In [60]:
# Write a function to clean up these example cases.
# We are shooting for uniformity because this will allow us to more easily recognize (non-exact) duplicates.
# We return the input after passing it through the .title() method which creates uniformity in case level.
# We also strip any whitespace using the .strip() method

def clean_title(job_title):
  if isinstance(job_title, str):
    return job_title.title().strip()
  else:
    return "Unknown"
  
for title in examples:
  print(clean_title(title))




Owner
Supervisor
Project Manager
Unknown


In [61]:
# Here we do a list comprehension

[clean_title(example) for example in examples]

['Owner', 'Supervisor', 'Project Manager', 'Unknown']

In [63]:
# Now we want to apply this function to the emp_title column...
# We can do a direct replacement

df['emp_title'] = df['emp_title'].apply(clean_title)

df.head()

loan_amnt  funded_amnt  ...  settlement_percentage settlement_term
0      20000        20000  ...                    NaN             NaN
1      25000        25000  ...                    NaN             NaN
2      26500        26500  ...                    NaN             NaN
3      10400        10400  ...                    NaN             NaN
4      35000        35000  ...                    NaN             NaN

[5 rows x 140 columns]

In [64]:
# Let's see how much progress we made.

df['emp_title'].value_counts(dropna=False, ascending=False)[:20]

Unknown               20947
Teacher                2557
Manager                2395
Registered Nurse       1418
Driver                 1258
Supervisor             1160
Truck Driver            920
Rn                      834
Office Manager          805
Sales                   803
General Manager         791
Project Manager         720
Owner                   625
Director                523
Operations Manager      518
Sales Manager           500
Police Officer          440
Nurse                   425
Technician              420
Engineer                412
Name: emp_title, dtype: int64

In [65]:
len(df['emp_title'].value_counts())

34902

In [66]:
# Null values?
df['emp_title'].isnull().sum()

0

In [0]:
# Originally there were 4,389 unique values...
# Now there is almost 1000 less values, simply by addressing capitalization and whitespace.
# And now, ZERO null values.

## PRACTICE WITH DATES

In [67]:
[col for col in df if col.endswith('_d')]

['issue_d', 'last_pymnt_d', 'next_pymnt_d', 'last_credit_pull_d']

In [68]:
df['issue_d'][:10]

0    Dec-2018
1    Dec-2018
2    Dec-2018
3    Dec-2018
4    Dec-2018
5    Dec-2018
6    Dec-2018
7    Dec-2018
8    Dec-2018
9    Dec-2018
Name: issue_d, dtype: object

In [70]:
df['issue_d'] = pd.to_datetime(df['issue_d'], infer_datetime_format=True)

df.dtypes[:15]

loan_amnt                       int64
funded_amnt                     int64
funded_amnt_inv               float64
term                           object
int_rate                      float64
installment                   float64
grade                          object
sub_grade                      object
emp_title                      object
emp_length                     object
home_ownership                 object
annual_inc                    float64
verification_status            object
issue_d                datetime64[ns]
loan_status                    object
dtype: object

In [71]:
df['issue_d'].iloc[0]

Timestamp('2018-12-01 00:00:00')

In [72]:
df['issue_d'].head().values

array(['2018-12-01T00:00:00.000000000', '2018-12-01T00:00:00.000000000',
       '2018-12-01T00:00:00.000000000', '2018-12-01T00:00:00.000000000',
       '2018-12-01T00:00:00.000000000'], dtype='datetime64[ns]')

In [73]:
df['issue_d'].dt.year

0         2018
1         2018
2         2018
3         2018
4         2018
          ... 
128407    2018
128408    2018
128409    2018
128410    2018
128411    2018
Name: issue_d, Length: 128412, dtype: int64

In [74]:
df['issue_d'].dt.month

0         12
1         12
2         12
3         12
4         12
          ..
128407    10
128408    10
128409    10
128410    10
128411    10
Name: issue_d, Length: 128412, dtype: int64

In [76]:
df['issue_year'] = df['issue_d'].dt.year
df['issue_month'] = df['issue_d'].dt.month

df.head()

loan_amnt  funded_amnt  ...  issue_year issue_month
0      20000        20000  ...        2018          12
1      25000        25000  ...        2018          12
2      26500        26500  ...        2018          12
3      10400        10400  ...        2018          12
4      35000        35000  ...        2018          12

[5 rows x 142 columns]

In [77]:
df['earliest_cr_line'].head()

0    Jun-1979
1    Sep-2003
2    Dec-2006
3    Jul-2013
4    Sep-1985
Name: earliest_cr_line, dtype: object

In [0]:
df['earliest_cr_line'] = pd.to_datetime(df['earliest_cr_line'], infer_datetime_format=True)

In [81]:
df['issue_d'] - df['earliest_cr_line']

0        14428 days
1         5570 days
2         4383 days
3         1979 days
4        12144 days
            ...    
128407   12114 days
128408   13241 days
128409    4505 days
128410    3652 days
128411    4413 days
Length: 128412, dtype: timedelta64[ns]

In [0]:
df['days_from_earliest_credit_to_issue'] = (df['issue_d'] - df['earliest_cr_line']).dt.days

In [83]:
df.head()

loan_amnt  funded_amnt  ...  issue_month days_from_earliest_credit_to_issue
0      20000        20000  ...           12                              14428
1      25000        25000  ...           12                               5570
2      26500        26500  ...           12                               4383
3      10400        10400  ...           12                               1979
4      35000        35000  ...           12                              12144

[5 rows x 143 columns]

In [84]:
df['days_from_earliest_credit_to_issue'].describe()

count    128412.000000
mean       5859.891490
std        2886.535578
min        1126.000000
25%        4049.000000
50%        5266.000000
75%        7244.000000
max       25171.000000
Name: days_from_earliest_credit_to_issue, dtype: float64

In [85]:
# Avg. credit history length
5859/365

16.052054794520547

In [86]:
25171/365

68.96164383561644

# Stretch Goals

You can do more with the LendingClub or Instacart datasets.

LendingClub options:
- There's one other column in the dataframe with percent signs. Remove them and convert to floats. You'll need to handle missing values.
- Modify the `emp_title` column to replace titles with 'Other' if the title is not in the top 20. 
- Take initiatve and work on your own ideas!

Instacart options:
- Read [Instacart Market Basket Analysis, Winner's Interview: 2nd place, Kazuki Onodera](http://blog.kaggle.com/2017/09/21/instacart-market-basket-analysis-winners-interview-2nd-place-kazuki-onodera/), especially the **Feature Engineering** section. (Can you choose one feature from his bulleted lists, and try to engineer it with pandas code?)
- Read and replicate parts of [Simple Exploration Notebook - Instacart](https://www.kaggle.com/sudalairajkumar/simple-exploration-notebook-instacart). (It's the Python Notebook with the most upvotes for this Kaggle competition.)
- Take initiative and work on your own ideas!

You can uncomment and run the cells below to re-download and extract the Instacart data

In [0]:
# !wget https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz

In [0]:
# !tar --gunzip --extract --verbose --file=instacart_online_grocery_shopping_2017_05_01.tar.gz

In [0]:
# %cd instacart_2017_05_01